### PR TITLE
init animations

### LINKED ISSUE #122: Animated Avatars
As a user avatars should have walk/jump/crouch/turn/etc animations when moving around. The animations should automatically blend from one to the next as your velocity and input changes.

### LINKED ISSUE #149: tickrate-independent camera-look & animation speed
deltaTime was not being applied to FirstPersonCamera & Animations (#122)

### PR DESCRIPTION
**Pull request by NateTheGreatt**
added
- [x] threejs animation integration
- [x] skinned mesh resource
- [x] animation system
- [x] animation mixing/blending
- [x] threejs scene graph gltf parsing
- [x] animation gltf parsing
- [x] audio limiter
- [x] physics event drain middleware
- [x] changed pickup from raycast to shapecast
- [x] delta time integration for camera & animations

velocity-based animations added
- [x] idle
- [x] turn in place
- [x] walk
  - [x] backwards
  - [x] strafe
- [x] run
  - [x] backwards
  - [x] strafe
- [x] jump/fall

issues
- [x] fix skinned mesh display issues
- [x] fix odd animation bug when blending between turning/rotation animations

resolves #122 
resolves #149

### CODE REVIEW

File: src/engine/animation/Skeleton.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
+export {};
```
Comment (robertlong) [inline src/engine/animation/Skeleton.ts]: Do we still need this file?

File: src/engine/animation/animation.game.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+export const AnimationComponent = new Map<number, IAnimationComponent>();
+export const BoneComponent = new Map<number, Bone>();
+
+const animationQuery = defineQuery([AnimationComponent]);
+const enterAnimationQuery = enterQuery(animationQuery);
+// const exitAnimationQuery = exitQuery(animationQuery);
+const boneQuery = defineQuery([BoneComponent]);
+
+/*
```
Comment (robertlong) [inline src/engine/animation/animation.game.ts:43]: Can you move this comment closer to where it's being used?

File: src/engine/animation/animation.game.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+  Run = "Run",
+  RunBack = "RunBack",
+}
+
+export const AnimationComponent = new Map<number, IAnimationComponent>();
+export const BoneComponent = new Map<number, Bone>();
+
+const animationQuery = defineQuery([AnimationComponent]);
+const enterAnimationQuery = enterQuery(animationQuery);
+// const exitAnimationQuery = exitQuery(animationQuery);
```
Comment (robertlong) [inline src/engine/animation/animation.game.ts:40]: Can we remove this?

File: src/engine/animation/animation.game.ts [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+function initializeAnimations(ctx: GameState) {
+  const entered = enterAnimationQuery(ctx.world);
+  for (let i = 0; i < entered.length; i++) {
+    const eid = entered[i];
+    const animation = AnimationComponent.get(eid);
+
+    if (animation) {
+      animation.actions = animation.clips.reduce((map, clip) => {
+        const action = animation.mixer.clipAction(clip).play();
```
Comment (robertlong) [inline src/engine/animation/animation.game.ts:117]: Should we still be playing animations automatically?

File: src/engine/audio/audio.main.ts [hunk 5]
```diff
@@ -87,16 +88,22 @@ export const AudioModule = defineModule<IMainThreadContext, MainAudioModule>({
...
 
     // hack - must play something thru the audio context for media streams to activate
-    (() => {
-      const osc = audioContext.createOscillator();
-      osc.frequency.value = 0;
-      osc.connect(audioContext.destination);
-      osc.start();
-      osc.stop();
-    })();
+    const osc = audioContext.createOscillator();
```
Comment (robertlong) [inline src/engine/audio/audio.main.ts:91]: This seems unrelated to this PR, it'd generally be better to split this out, but it's ok for now. What is this doing though?

File: src/engine/node/node.render.ts [hunk 6]
```diff
@@ -143,6 +169,10 @@ export function updateLocalNodeResources(
         node.mesh = undefined;
       }
 
+      // TODO: skinned mesh disposal
```
Comment (robertlong) [inline src/engine/node/node.render.ts:172]: Hmm, yeah we need to look into what needs to be disposed here. If you don't do it in this PR make sure to create an issue so that we look into it soon.

File: src/engine/physics/physics.game.ts [hunk 7]
```diff
@@ -130,6 +121,18 @@ export const PhysicsSystem = (state: GameState) => {
     }
   }
 
+  // remove rigidbody from physics world
+  const exited = exitedPhysicsQuery(world);
```
Comment (robertlong) [inline src/engine/physics/physics.game.ts:125]: Does this fix our issue with rigid body disposal? Should it close https://github.com/matrix-org/thirdroom/issues/147 ?

File: src/plugins/CubeSpawner.ts [hunk 8]
```diff
@@ -173,9 +202,8 @@ const _impulse = new RAPIER.Vector3(0, 0, 0);
 
 const cameraWorldQuat = quat.create();
 export const CubeSpawnerSystem = (ctx: GameState) => {
-  const module = getModule(ctx, CubeSpawnerModule);
+  // const module = getModule(ctx, CubeSpawnerModule);
```
Comment (robertlong) [inline src/plugins/CubeSpawner.ts:205]: This can be removed

**Comment by robertlong** [comment]
I found one small bug when strafing backwards (Holding A and S)
https://user-images.githubusercontent.com/1753624/183532647-205a4a09-54eb-4837-bc83-6126e750e60f.mov

File: src/engine/audio/audio.main.ts (hunk 5 above)
Comment (NateTheGreatt) [reply src/engine/audio/audio.main.ts:91]: it's a limiter with a bit of headroom to protect speakers from spikes in amplitude and take the edge off of hard clipping from the decibel ceiling when audio gets too loud/chaotic

File: src/engine/physics/physics.game.ts (hunk 7 above)
Comment (NateTheGreatt) [reply src/engine/physics/physics.game.ts:125]: i thought it did, but it did not unfortunately :(

Comment (NateTheGreatt) [reply src/engine/physics/physics.game.ts:125]: i added notes to #147

File: src/engine/animation/animation.game.ts (hunk 4 above)
Comment (NateTheGreatt) [reply src/engine/animation/animation.game.ts:95]: yes, but they don't need to be `enabled` all the time. i added appropriate enables/disables, should save the mixer a few cycles

**Comment by robertlong** [comment]
And this one which is due to frustum culling. If you shrink the bounding box down I think it'll be fine.

https://user-images.githubusercontent.com/1753624/183533171-a6ffe60a-b7ec-411c-b121-4ebf723a0c0d.mov

**Comment by NateTheGreatt** [comment]
> I found one small bug when strafing backwards (Holding A and S) https://user-images.githubusercontent.com/1753624/183532647-205a4a09-54eb-4837-bc83-6126e750e60f.mov

fixed!

**Comment by robertlong** [comment]
Awesome work!
